### ISSUE TITLE
Draftail fails to turn images into hyperlinks

### ISSUE THREAD

**Issue by jponce-tz** [wagtail/wagtail#7731]
### Issue Summary

Draftail fails to turn images into hyperlinks.

### Steps to Reproduce

1. Create a new page.
2. Insert a new image.
3. Select/highlight the image and press on the "link" button in the editor.
4. The editor will show a "Server error" toast notification and the logs show the following error:
```
[23/Nov/2021 09:51:38] "GET /admin/choose-page/undefined/?page_type=wagtailcore.page&allow_external_link=true&allow_email_link=true&allow_phone_link=true&allow_anchor_link=true&link_text= HTTP/1.1" 404 9096
Not Found: /admin/choose-page/undefined/ 
```

I was expecting to have the possibility to turn the image into a hyperlink, either using the link button in the editor, or having an URL field when inserting the image. In any case, trying to turn the image into a link should not give that error.

* I have confirmed that this issue can be reproduced as described on a fresh Wagtail project: yes

### Technical details

* Python version: 3.9.5
* Django version: 3.2.9
* Wagtail version: 2.15.1
* Browser version: Chrome 95

**Comment by thibaudcolas** [comment]
Hi @jponce-tz, this isn’t supported at the moment. Links and images are different, separate entities in rich text, and we’d need a third concept of "image with link" to support this. I’d recommend using StreamField for this at the moment.

I’m tempted to mark this as a bug because I would have expected the editor to support adding a link, thereby removing the image in a case like this – however it’s not too clear to me that’s much more desirable than the current behavior.

**Comment by realjumy** [comment]
Hi @thibaudcolas . Thanks for taking a look into it.

At the moment the editor throws an error when an user attempts to assign a link to an image, so it would be great if at least that is sorted.

Regarding the use of StreamField... I agree, that would be the right approach. I like the way StreamField works and and it is always my preferred option in all my projects. However, there are certain situations where it is not an option, and the users expect that assigning hyperlink to images (something that is trivial in many other editors) is a possibility.

How difficult is adding that "image with link" concept?

**Comment by jponce-tz** [comment]
Hi @thibaudcolas . Thanks for taking a look into it.

At the moment the editor throws an error when an user attempts to assign a link to an image, so it would be great if at least that is sorted.

Regarding the use of StreamField... I agree, that would be the right approach. I like the way StreamField works and and it is always my preferred option in all my projects. However, there are certain situations where it is not an option, and the users expect that assigning hyperlink to images (something that is trivial in many other editors) is a possibility.

How difficult is adding that "image with link" concept?

PS. Sorry if you were notified twice about this message. I initially answered with a different account and I deleted it and sent it again with the right one for the sake of avoiding confusion.

**Comment by thibaudcolas** [comment]
@jponce-tz I don’t think it would be too complex, but considering this is normally what we encourage using StreamField for (since it’s much more flexible), there isn’t a lot of incentive to implement this in rich text. Can you provide further details on which situations this isn’t an option in?

**Comment by jponce-tz** [comment]
In our particular case is not an option because we have some clients that are opposed to StreamField. For some reason they dislike it and they specifically ask for a "classic editor". Maybe they have difficulties understanding how it works, or maybe they are from a different generation and they want something that is similar to what they have been always using. Whatever the reason, we even had to invest quite a lot of time in trying to adapt ckeditor5 for Wagtail, but amount of time that it required for adapting it and maintaining it was simply too much for the team. A part from that, we have no one complaining about anything else from Wagtail. They are actually glad to use something different to Wordpress or Drupal for a change.

And I absolutely agree with you: StreamField is more flexible and I like the concept the better. And I don't see any difficulty in using it, but some people... I guess they struggle to grasp the way it works.

**Comment by def324** [comment]
We had a similar issue and ended up using a custom image model to achieve this. It's well documented in the following sections and pretty easy to customize (just change the caption logic to a link).

- https://docs.wagtail.org/en/stable/advanced_topics/images/custom_image_model.html
- https://docs.wagtail.org/en/stable/advanced_topics/images/changing_rich_text_representation.html

It took a bit of work and required a data migration for our existing images but in the end it worked out for us. The only downside is that one image can have only one link assigned. So if you want the same image to link to different pages in different parts of the site you'd have to upload the same image multiple times. 

Hope someone finds this helpful :)

**Comment by chano195** [comment]
Try migrating from richtext to Streamfields to be able to create a link with an image inside. I achieved it. but another problem arose. How am I supposed to implement the alignment to the images? Now users will not be able to align the images, I don't know what they mean when they say that the streamfields are more flexible, but in this case it is not at all. or at least I couldn't figure out how to align the images with links other than by creating 3 different blocks (image aligned to the right, image aligned to the left, full width image)._.
I think this is just a small example of the limitations.

**Comment by chano195** [comment]
@gasman Thank you for responding so quickly, I think I was the one who was not very flexible haha. I don't have much experience with StreamFields. I have been trying to implement ChoiceBlock without success. I think I'm doing something wrong. If you can help me understand what I'm doing wrong I would be very grateful, this is my code:
"
class ImagePlusLinkWide(blocks.StructBlock):
     link= blocks.URLBlock()
     image= ImageChooserBlock()

class ImagePlusLinkLeft(blocks.StructBlock):
     link= blocks.URLBlock()
     image= ImageChooserBlock()

class ImagePlusLink(blocks.StructBlock):
     image= blocks.ChoiceBlock(
         ('wide', ImagePlusLinkWide()),
         ('left', ImagePlusLinkLeft()),
     )

class Content(ClusterableModel,Orderable):
     body = StreamField([
         ('text', blocks.RichTextBlock(label='Rich Text')),
         ('image_plus_link',ImagePlusLink([
         ], icon='image', label='Image with link')),
     ], blank=True)
"

**Comment by asset-web** [comment]
> [@jponce-tz](https://github.com/jponce-tz) I don’t think it would be too complex, but considering this is normally what we encourage using StreamField for (since it’s much more flexible), there isn’t a lot of incentive to implement this in rich text. Can you provide further details on which situations this isn’t an option in?

Can someone expand upon how this is implemented in the StreamField?  Currently, I am using a StreamField field and it includes the use of ImageChooserBlock and an error is also thrown for me too, when a user tries to add a link to an image.  Or was that reply that a custom block needs to be developed for StreamField for any Wagtail application wanting to support end users adding links to images?

Error shown in the Javascript console:
`Failed to load resource: the server responded with a status of 404` ref: `/cms/choose-page/undefined/?page_type=wagtailcore.page&allow_external_link=true&allow_email_link=true&allow_phone_link=true&allow_anchor_link=true&link_text=`

Python version: 3.13.9
Django version: 5.2.9
Wagtail version: 6.3.5
Browser version: Chrome 143

**Comment by asset-web** [comment]
In case of use to anyone else [Image Links](https://docs.coderedcorp.com/wagtail-crx/features/contentblocks/imagelink.html) from these set of extensions to Wagtail CMS

[Wagtail + CodeRed Extensions](https://docs.coderedcorp.com/wagtail-crx/index.html)

**Comment by jponce-tz** [comment]
As @asset-web mentioned, I think the best option is to use CodeRed Extensions.
